File: 2465-shifting-letters-ii/2465-shifting-letters-ii.c
char* shiftingLetters(char* s, int** shifts, int shiftsSize, int* shiftsColSize) {
    int n = strlen(s);
    int prefix[n + 1] = {};

    for (uint i = 0; i < shiftsSize; i++) {
        int s = shifts[i][0];
        int e = shifts[i][1];
        int d = shifts[i][2];

        prefix[s] += (2 * d - 1);
        prefix[e + 1] -= (2 * d - 1);
    }

    int curr = 0;
    for (int i = 0; i < n; i++) {
        curr = (curr + prefix[i]) % 26;
        s[i] = 'a' + (s[i] - 'a' + curr + 26) % 26;
    }
    return s;
}

File: 2465-shifting-letters-ii/2465-shifting-letters-ii.js
/**
 * @param {string} s
 * @param {number[][]} shifts
 * @return {string}
 */
var shiftingLetters = function (s, shifts) {
    const n = s.length;
    const prefix = new Array(n + 1).fill(0);

    for (let shift of shifts) {
        const [st, ed, dr] = shift;
        prefix[st] += (2 * dr - 1);
        if (ed + 1 < n) {
            prefix[ed + 1] -= (2 * dr - 1);
        }
    }

    let sb = [];
    let currShift = 0;

    for (let i = 0; i < n; i++) {
        currShift = (currShift + prefix[i]) % 26;
        const ch = String.fromCharCode(((s.charCodeAt(i) - 'a'.charCodeAt(0) + currShift + 26) % 26) + 'a'.charCodeAt(0));
        sb.push(ch);
    }

    return sb.join('');

};